### PR TITLE
Week 4: TDD + Scheduling Algorithm

### PR DESCRIPTION
**Pull request by riku1212**
TDD Intro:
- modify code to trim whitespace and remove special characters such as `@%#$`
- add relevant test to test the functionality of symbol removal

TDD Project:
- add Claudia's test case to the test file
- Algorithm: Let E be the number of events running and P be the maximum number of participants in each event. Then, the algorithm will run in O(ElogE + EP) time. Firstly, we need to create `overlappingTime`, which takes O(EP) as we need to iterate through each events and use `retainAll` to check for intersections (takes O(P) time). After that, we need to merge this array, which can be done by sorting it (takes O(ElogE) time) and merging it greedily, which takes O(E) time as we only need to iterate and all operations within are O(1). Finally, we can use sliding window to mark available time which satisfy the duration. Again, this takes O(E) time. So, the total complexity is O(ElogE + EP). 

Note that when E is big (>200), we can construct a better algorithm with complexity O(EP + N) where N is the number of timestamps. One modification we can do is use a Boolean Array with size N (which is finite!), in this case the size is only 24*60=1440 instead of sorting the events. We first initialize this array with `true`. After determining which timestamps are unavailable, we immediately mark them as `false` inside this Boolean array. After that, we can just iterate through this array to determine which timestamp is available and compare them accordingly with the intended duration.

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,96 @@
...
+    * MeetingRequest. A TimeRange is available if the person in MeetingRequest are all available.
+    * We first do this by checking which TimeRange are the attendees not available in and 
+    * use sliding window to record the available TimeRange provided they satisfy the duration.
+    * @param    events  a collection of Event objects
+    * @param    request a MeetingRequest object
+    * @return           collection of available TimeRange  
+    */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        ArrayList<TimeRange> overlappingTime = new ArrayList<TimeRange>();
+        ArrayList<TimeRange> mergedEventTime = new ArrayList<TimeRange>();
```
Comment (kai-dream) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Nitpick: No need to initialize mergedEventTime as you assigned mergedEventTime below.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,96 @@
...
+                overlappingTime.add(event.getWhen());
+            }
+        }
+
+        // merge eventTime time range
+        mergedEventTime = mergeEventTime(overlappingTime);
+
+        // sliding window to find available timing that matches the duration
+        int prevEnd = TimeRange.START_OF_DAY;
+        long duration=request.getDuration();
```
Comment (kai-dream) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Minor style inconsistency: leave one space before and after '='. That's basically the same for all binary operators.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,96 @@
...
+    * overlapping. 
+    * @param    events  a collection of Event objects
+    * @return           collection of merged event time, sorted by start
+    */
+    private ArrayList<TimeRange> mergeEventTime(ArrayList<TimeRange> eventTime){
+        if (eventTime.size() < 1) {return eventTime;}
+
+        Collections.sort(eventTime, TimeRange.ORDER_BY_START);
+        ArrayList<TimeRange> mergedEventTime = new ArrayList();  
+        int prevStart = eventTime.get(0).start(), prevEnd = eventTime.get(0).end();
```
Comment (kai-dream) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good naming of variable names here.

A small suggestion in style, separating the two variable initialisation into two lines.
Usually the only situation we initialise multiple variables in one single line is when they are super tightly coupled, and initialised only using simple values. For example, (x, y, z) when they represent the 3 component of a coordinate. Benefits in doing so:
* you can put breakpoints in each line.
* It is easier to visually read, compare, and tell the difference in initialising each variable.

**Review by kai-dream — APPROVED** [review]
Great algorithm.

Time complexity analysis can actually be part of the function comment instead of in the PR description. 

I think you missed to consider the complexity of sorting, so you might want to add that.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (riku1212) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ok noted!

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (riku1212) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ah i missed that. what would one do to get a linter :)

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (riku1212) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: ah i see. usually i make them a one-liner if they have the same type to reduce line count. but i guess breaking them will improve readability if the initialization is too complex. thanks!

**Comment by kai-dream** [comment]
> Great algorithm.
> 
> Time complexity analysis can actually be part of the function comment instead of in the PR description.
> 
> I think you missed to consider the complexity of sorting, so you might want to add that.

My bad, you already consider sorting in the ELogE component. All good.

**Review by kai-dream — APPROVED** [review]
LGTM +1

Thanks for integrating my feedback!

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (kai-dream) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks. Just a suggestion, reducing line count is usually the least important factor to balance, in terms of weight when considering different factors, I recommend this order: the ability to place breakpoints > readability >>> line count. 

For example sometimes it is perfectly fine to put 3 blank lines just to make your code structurally "looks" like your logical structure.

**Pull request merged**